### PR TITLE
[gui] Implemented movingspeed to mover/resize controls

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Since a new class has been added to simulate the speed of movement `CMovingSpeed`, i have replaced the old implementation,
this implementation now allows much more control and can be configured through skin XML 

This will fix also the bug in Calibration window that when you move the pixel square, in some cases it is not possible to restore the value to 0. This is because pixels were considered with decimal values instead of integers.

### `movingspeed` tag
- Is the parent tag of the feature
- Can contains optionals global attribute values, that will be applied on each event type, unless otherwise specified in the children themselves
- Attributes supported: acceleration, maxvelocity, resettimeout, delta

### `eventconfig` tag
- Is the only child supported by `movingspeed` tag
- Attributes supported: type, acceleration, maxvelocity, resettimeout, delta
- Mandatory attribute is `type` 

### Tag attributes
`type`: The direction of movement, currently accepted values are: `up`, `down`, `left`, `right`
`acceleration`: Acceleration in pixels per second (px/sec) (int type)
`maxvelocity`: Max movement speed, it depends from acceleration value a suitable value could be (acceleration value)*3. Set 0 to disable it. (int type)
`resettimeout`: Will resets the acceleration speed if idle for specified millisecs (int type)
`delta`: Specify the minimal increment step can be as integer of float (for example for pixel movement we have to increment by 1 without decimals) (float or int type)

### Example 1
Example of global configuration applied to all moviments types except one (override global config):
```xml
<control type="mover" id="8">
...
	<movingspeed acceleration="180" maxvelocity="300" resettimeout="200" delta="1">
		<eventconfig type="up" />
		<eventconfig type="down" />
		<eventconfig type="left" acceleration="100" maxvelocity="100" resettimeout="160" delta="1">
		<eventconfig type="right" />
	</movingspeed>
</control>
```
In this case the control can be moved in 4 directions

### Example 2
Set eventconfig without some attributes

```xml
<control type="mover" id="8">
...
	<movingspeed acceleration="180" maxvelocity="300" resettimeout="200" delta="1">
		<eventconfig type="left" acceleration="100" maxvelocity="100">
	</movingspeed>
</control>
```

These attributes `resettimeout` and `delta` since are not specified in `eventconfig` tag, will be taken from the global (if sets).
In this case the control can be moved in 1 direction only

## Motivation and context

Mover and resize controls had incomplete implementation of speed simulation

## How has this been tested?

Calibration window is the only xml that use these controls

## What is the effect on users?

none

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/MovingSpeed.h [hunk 1]
```diff
@@ -28,11 +30,39 @@ struct EventCfg
...
+  {
+  }
+
   float m_acceleration;
   float m_maxVelocity;
   uint32_t m_resetTimeout;
+  float m_delta{0};
 };
 
+enum class EventType : uint32_t
```
Comment (enen92) [inline xbmc/utils/MovingSpeed.h:55]: `: uint32_t` is likely not needed here

**Review by enen92 — COMMENTED** [review]
This needs update on the doxyfile for the mover control: https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIMoverControl.dox

**Comment by CastagnaIT** [comment]
> This needs update on the doxyfile for the mover control: https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIMoverControl.dox

you have right
I knew something was missing but i couldn't remember where...

**Comment by CastagnaIT** [comment]
@ProfYaffle please if you have time can you review for language spelling typo mainly on:
GUIMoverControl.dox
GUIResizeControl.dox
thanks

File: xbmc/guilib/GUIMoverControl.dox [hunk 2]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
 |----------------:|:--------------------------------------------------------------|
 | texturefocus    | Specifies the image file which should be displayed when the mover has focus. [See here for additional information about textures](http://kodi.wiki/view/Texture_Attributes).
 | texturenofocus  | Specifies the image file which should be displayed when the mover does not have focus.
+| movingspeed     | Specifies the allowed directional movements and their configuration. The <c>`<movingspeed>`</c> must contains as many <c>`<eventconfig>`</c> tags as required.
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIMoverControl.dox:46]: Specifies the allowed directional movements and respective configurations. The <c>`<movingspeed>`</c> can contain as many <c>`<eventconfig>`</c> tags as required.

File: xbmc/guilib/GUIMoverControl.dox [hunk 3]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
...
+
+The <c>`movingspeed`</c> tag must be specified to allow the control to be moved via
+an input device such as keyboard. Each direction can be specified and configured
+to simulate the motion effect with the <c>`eventconfig`</c> tag.
+
+The following attributes allow the motion effect of the control to be configured:
+
+| Attribute       | Description                                                   |
+|----------------:|:--------------------------------------------------------------|
+| type            | Specifies the direction of movement, accepted values are: <c>`up`</c>, <c>`down`</c>, <c>`left`</c>, <c>`right`</c>, all types of direction are optional.
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIMoverControl.dox:60]: Specifies the direction of movement. Accepted values are: <c>`up`</c>, <c>`down`</c>, <c>`left`</c>, <c>`right`</c>. All directions are optional and do not have to be specified.

File: xbmc/guilib/GUIMoverControl.dox [hunk 4]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
...
+| acceleration    | Specifies the acceleration in pixels per second (integer value).
+| maxvelocity     | Specifies the max movement speed, it depends from acceleration value, a suitable value could be (acceleration value)*3. Set 0 to disable it. (integer value).
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIMoverControl.dox:62]: Specifies the maximum movement speed. This depends on the acceleration value, e.g. a suitable value could be (acceleration value)*3. Set to 0 to disable. (integer value).

File: xbmc/guilib/GUIMoverControl.dox [hunk 5]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
...
+| resettimeout    | Specifies the idle timeout to resets the acceleration speed in millisecs (integer value).
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIMoverControl.dox:63]: Specifies the idle timeout before resetting the acceleration speed (in milliseconds, integer value).

File: xbmc/guilib/GUIMoverControl.dox [hunk 6]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
...
+| delta           | Specifies the minimal pixel increment step (integer value).
+
+The attributes <c>`acceleration`</c>, <c>`maxvelocity`</c>, <c>`resettimeout`</c>,
+<c>`delta`</c> must be specified individually for each <c>`eventconfig`</c> tag,
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIMoverControl.dox:67]: can be specified individually

File: xbmc/guilib/GUIMoverControl.dox [hunk 7]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
...
+or globally in to <c>`movingspeed`</c> tag, or also mixed, as the following example:
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIMoverControl.dox:68]: or globally in the <c>`movingspeed`</c> tag, or a mixture, as the following example:

File: xbmc/guilib/GUIResizeControl.dox [hunk 8]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
 |----------------:|:--------------------------------------------------------------|
 | texturefocus    | Specifies the image file which should be displayed when the resizer has focus. [See here for additional information about textures](http://kodi.wiki/view/Texture_Attributes).
 | texturenofocus  | Specifies the image file which should be displayed when the resizer does not have focus.
+| movingspeed     | Specifies the allowed directional movements and their configuration. The <c>`<movingspeed>`</c> must contains as many <c>`<eventconfig>`</c> tags as required.
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIResizeControl.dox:47]: As above

File: xbmc/guilib/GUIResizeControl.dox [hunk 9]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
...
+
+The <c>`movingspeed`</c> tag must be specified to allow the control to be moved via
+an input device such as keyboard. Each direction can be specified and configured
+to simulate the motion effect with the <c>`eventconfig`</c> tag.
+
+The following attributes allow the motion effect of the control to be configured:
+
+| Attribute       | Description                                                   |
+|----------------:|:--------------------------------------------------------------|
+| type            | Specifies the direction of movement, accepted values are: <c>`up`</c>, <c>`down`</c>, <c>`left`</c>, <c>`right`</c>, all types of direction are optional.
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIResizeControl.dox:61]: As above

File: xbmc/guilib/GUIResizeControl.dox [hunk 10]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
...
+| acceleration    | Specifies the acceleration in pixels per second (integer value).
+| maxvelocity     | Specifies the max movement speed, it depends from acceleration value, a suitable value could be (acceleration value)*3. Set 0 to disable it. (integer value).
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIResizeControl.dox:63]: As above

File: xbmc/guilib/GUIResizeControl.dox [hunk 11]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
...
+| resettimeout    | Specifies the idle timeout to resets the acceleration speed in millisecs (integer value).
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIResizeControl.dox:64]: As above

File: xbmc/guilib/GUIResizeControl.dox [hunk 12]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
...
+| delta           | Specifies the minimal pixel increment step (integer value).
+
+The attributes <c>`acceleration`</c>, <c>`maxvelocity`</c>, <c>`resettimeout`</c>,
+<c>`delta`</c> must be specified individually for each <c>`eventconfig`</c> tag,
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIResizeControl.dox:68]: As above

File: xbmc/guilib/GUIResizeControl.dox [hunk 13]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
...
+or globally in to <c>`movingspeed`</c> tag, or also mixed, as the following example:
```
Comment (ProfYaffle) [inline xbmc/guilib/GUIResizeControl.dox:69]: As above

File: xbmc/guilib/GUIMoverControl.dox (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/guilib/GUIMoverControl.dox]: thank you this and all other changes applied

**Comment by CastagnaIT** [comment]
is needed to do other changes with this PR? need ask review from others?

File: xbmc/guilib/GUIMoverControl.dox [hunk 14]
```diff
@@ -37,10 +43,42 @@ important, as xml tags are case-sensitive.
 |----------------:|:--------------------------------------------------------------|
 | texturefocus    | Specifies the image file which should be displayed when the mover has focus. [See here for additional information about textures](http://kodi.wiki/view/Texture_Attributes).
 | texturenofocus  | Specifies the image file which should be displayed when the mover does not have focus.
+| movingspeed     | Specifies the allowed directional movements and respective configurations. The <c>`<movingspeed>`</c> tag section can contain as many <c>`<eventconfig>`</c> tags as required.
 
 
 --------------------------------------------------------------------------------
-\section Mover_Control_sect3 See also
+\section Resize_Control_sect3 Note on use of movingspeed tag
+
```
Comment (enen92) [inline xbmc/guilib/GUIMoverControl.dox:51]: Missing `@skinning_v20` revision tag somewhere

File: xbmc/guilib/GUIMoverControl.h [hunk 15]
```diff
@@ -15,27 +15,28 @@
 
 #include "GUIControl.h"
 #include "GUITexture.h"
+#include "utils/MovingSpeed.h"
 
 #define ALLOWED_DIRECTIONS_ALL   0
 #define ALLOWED_DIRECTIONS_UPDOWN  1
 #define ALLOWED_DIRECTIONS_LEFTRIGHT 2
 
-#define DIRECTION_NONE 0
```
Comment (enen92) [inline xbmc/guilib/GUIMoverControl.h:23]: Just a note, likely doesn't apply here but be careful with defines sometimes they get exposed to the python api

File: xbmc/guilib/GUIResizeControl.dox [hunk 16]
```diff
@@ -38,10 +44,42 @@ important, as `xml` tags are case-sensitive.
...
 
 --------------------------------------------------------------------------------
-\section Resize_Control_sect3 See also
+\section Resize_Control_sect3 Note on use of movingspeed tag
+
+The <c>`movingspeed`</c> tag must be specified to allow the control to be moved via
+an input device such as keyboard. Each direction can be specified and configured
+to simulate the motion effect with the <c>`eventconfig`</c> tag.
+
+The following attributes allow the motion effect of the control to be configured:
```
Comment (enen92) [inline xbmc/guilib/GUIResizeControl.dox:57]: Missing revision tag somewhere

**Review by enen92 — APPROVED** [review]
Looks good to me

File: xbmc/guilib/GUIMoverControl.dox (hunk 14 above)
Comment (CastagnaIT) [reply xbmc/guilib/GUIMoverControl.dox:51]: add a revision section as done in ranges control

File: xbmc/guilib/GUIMoverControl.h (hunk 15 above)
Comment (CastagnaIT) [reply xbmc/guilib/GUIMoverControl.h:23]: i tried to make a search but i havent found no reuse of this define

File: xbmc/guilib/GUIResizeControl.dox (hunk 16 above)
Comment (CastagnaIT) [reply xbmc/guilib/GUIResizeControl.dox:57]: done as above

**Comment by CastagnaIT** [comment]
jenkins build this please

**Comment by enen92** [comment]
@CastagnaIT if possible please try to add something on https://forum.kodi.tv/showthread.php?tid=363553&page=2 to let skinners know about the new feature

**Comment by CastagnaIT** [comment]
done

**Pull request merged**